### DISCUSSION TITLE
"Context-dependent expressions" in the `aes()`

### DISCUSSION THREAD

**Discussion by yjunechoe** [ggplot2-extenders/ggplot-extension-club#23] [Ideas]
Since `aes()` is data-masked, like `mutate()`, it's interesting to think about parallels to dplyr-style ["context-dependent expressions"](https://dplyr.tidyverse.org/reference/context.html) like `n()`.

From a discussion in https://github.com/tidyverse/ggplot2/issues/5505#issuecomment-1791324008, I proposed a (version of Teun's) `prop_by()` function for use in the `after_stat()` for layers with `StatCount`.

```r
prop_by <- function(...) {
  count <- evalq(count, parent.frame())
  if (length(list(...)) == 0) {
    count / sum(count)
  } else {
    count / ave(count, ..., FUN = function(x) sum(abs(x)))
  }
}
```

I'm circling back to this now because I've found myself using this function a lot recently for making stacked bar plots where, for a second, GeomLabel layer, I want to show proportions by x (the intuitive grouping of proportions for a stacked bar plot), rather by panel as a whole

```r
ggplot(diamonds, aes(cut, group = color)) +
  geom_bar(
    aes(fill = color),
    position = position_fill()
  ) +
  geom_label(
    aes(label = after_stat(scales::percent(prop_by(x), 2))),
    position = position_fill(vjust = 0.5),
    stat = "count"
  )
```

![image](https://github.com/teunbrand/ggplot-extension-club/assets/52832839/fa8c10b4-1911-43df-b8f7-98734d14c5eb)

Over-engineering this a little bit with a slight hack, you can also check whether `prop_by()` is being called in the appropriate environment:
- Is called inside `after_stat()`:
  ```r
  any(sapply(sys.calls(), rlang::is_call, "after_stat"))
  ```
- Is used in a layer with `StatCount` (this could also just loosely check for `count` column present in after-stat data)
  ```r
  identical(sys.frames()[[sapply(sys.calls(), rlang::is_call, "map_statistic")]]$self$stat, StatCount)
  ```
  
---

So far I can't think of any pairing other than `StatCount` + `prop_by()` that actually has potential to be practical/useful, but maybe that just reflects the limits of _my_ creativity 🤣. Anyways, if we find a critical mass of `prop_by()`-esque functions, I think it'd be fun to write an extension package with various extensions at this interface (you wouldn't even have to import ggplot!).

**Comment by yjunechoe** [comment]
Also TIL: if you want to invoke the debugger via `browser()` in the after stat context, this obvious attempt fails to preserve the call stack

```r
ggplot(diamonds, aes(cut)) +
  geom_bar(aes(y = after_stat(browser())))
# Browse[1]> length(sys.calls())
# [1] 0
```

But if you call `browser()` as an immediately invoked function expression, you preserve what came before it's called in the internals:

```r
ggplot(diamonds, aes(cut)) +
  geom_bar(aes(y = after_stat((\()browser())())))
# Browse[1]> length(sys.calls())
# [1] 18
```

**Reply by teunbrand** [reply]
There is definitely some voodoo/witchcraft at work with evaluating the `after_stat()` content, but why the call stack is empty in the first case and not the second beats me 🤷

**Comment by teunbrand** [comment]
TIL you can sort-of reconsistute the data from inside an expression in `aes()`.

``` r
do_by <- function(expr, ...) {
  expr <- rlang::enquo(expr)
  data <- as.data.frame(as.list(parent.frame()$.top_env))
  groups <- split(data, interaction(...))
  answer <- vector("list", length(groups))
  for (i in seq_along(groups)) {
    answer[[i]] <- rlang::eval_tidy(expr, groups[[i]])
  }
  unlist(answer)
}

library(ggplot2)
ggplot(mpg, aes(displ, hwy)) +
  geom_point(aes(colour = do_by(cty / mean(cty), drv)))
```

![](https://i.imgur.com/cGnHRTT.png)

<sup>Created on 2024-06-22 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by yjunechoe** [reply]
TIL `.top_env` 🤯

Looks like it's just the column order and dataframe-level attributes that can't be recovered? Good enough for me!

**Reply by teunbrand** [reply]
I think `.top_env` is a tidy eval specific thing and probably wouldn't work anywhere else

**Reply by yjunechoe** [reply]
So it seems like `.top_env` is kinda like a more explicit version of the `.data` pronoun (?) I've always wondered how to "get" all the contents of `.data` so this is useful to know 😁

**Reply by teunbrand** [reply]
I'd bet this is just an implementation detail and not something wilfully exposed to users, but it is fun to find these types of details

**Comment by larmarange** [comment]
Just to mention that it exists an alternative approach using a dedicated stat, cf. https://larmarange.github.io/ggstats/articles/stat_prop.html

**Reply by EvaMaeRey** [reply]
The 'by' aes seems intuitive to me! (I've called it 'within' elsewhere, but I think I kind of favor 'prop_by' or 'prop_within' now) I think it's nice to have something like this in cases we can expect the proportion computation to be frequently of interest.

**Comment by EvaMaeRey** [comment]
Another context for prop_by @yjunechoe!
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/a17fc0b2-ca28-4668-bd5c-0e00917b7791)

To percents (globally)

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/b389e095-00ee-4b63-8008-f59c06f2bb88)

To percents (within contintent)

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/a710b70b-1ddd-4434-8fd2-f978ad06ba1d)


But I think I kind of favor building a prop column (by defining scope via 'within' or 'by' aes)  in the stat because it does seem like a routine kind of task.
